src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs

```diff
@@ -0,0 +1,895 @@
+            var source = @"
+using System;
+using System.Collections.Generic;
+
+class Program
+{
+    static void Main(string[] args)
+    {
+        int Local(this int x)
+        {
+            return x;
+        }
+        Console.WriteLine(Local(2));
+    }
+}
+";
+            var option = TestOptions.ReleaseExe.WithWarningLevel(0);
+            CreateCompilationWithMscorlibAndSystemCore(source, options: option, parseOptions: _parseOptions).VerifyDiagnostics(
+    // (9,13): error CS1106: Extension method must be defined in a non-generic static class
+    //         int Local(this int x)
+    Diagnostic(ErrorCode.ERR_BadExtensionAgg, "Local").WithLocation(9, 13)
+                );
+        }
+
+        [Fact]
+        public void InferredReturn()
+        {
+            var source = @"
+using System;
+using System.Collections.Generic;
+
+class Program
+{
+    static void Main(string[] args)
+    {
+        var Local()
+        {
+            return 2;
+        }
+        Console.WriteLine(Local());
+    }
+}
+";
+            // message is temporary
+            var option = TestOptions.ReleaseExe.WithWarningLevel(0);
+            CreateCompilationWithMscorlibAndSystemCore(source, options: option, parseOptions: _parseOptions).VerifyDiagnostics(
+    // (9,9): error CS7019: Type of 'Program.Local()' cannot be inferred since its initializer directly or indirectly refers to the definition.
+    //         var Local()
+    Diagnostic(ErrorCode.ERR_RecursivelyTypedVariable, "var").WithArguments("Program.Local()").WithLocation(9, 9)
+                );
+        }
+
+        [Fact]
+        public void ArglistIterator()
+        {
+            var source = @"
+using System;
+using System.Collections.Generic;
+
+class Program
+{
+    static void Main(string[] args)
+    {
+        IEnumerable<int> Local(__arglist)
+        {
+            yield return 2;
+        }
+        Console.WriteLine(string.Join("","", Local(__arglist())));
+    }
+}
+";
+            var option = TestOptions.ReleaseExe.WithWarningLevel(0);
+            CreateCompilationWithMscorlibAndSystemCore(source, options: option, parseOptions: _parseOptions).VerifyDiagnostics(
+    // (9,26): error CS1636: __arglist is not allowed in the parameter list of iterators
+    //         IEnumerable<int> Local(__arglist)
+    Diagnostic(ErrorCode.ERR_VarargsIterator, "Local").WithLocation(9, 26)
+                );
+        }
+
+        [Fact]
+        public void ForwardReference()
+        {
+            var source = @"
+using System;
+using System.Collections.Generic;
+
+class Program
+{
+    static void Main(string[] args)
+    {
+        Console.WriteLine(Local());
+        int Local() => 2;
+    }
+}
+";
+            var option = TestOptions.ReleaseExe.WithWarningLevel(0);
+            CreateCompilationWithMscorlibAndSystemCore(source, options: option, parseOptions: _parseOptions).VerifyDiagnostics(
+    // (9,27): error CS0841: Cannot use local variable 'Local' before it is declared
```

Comment:
Same thing as above - we can add a new error ID and message later on, when things are a bit more concrete.